temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java

```diff
@@ -304,7 +309,10 @@ private void handleEventsBatch(List<HistoryEvent> events, boolean hasNextEvent)
       }
 
       try {
-        handleSingleEvent(event, iterator.hasNext() || hasNextEvent);
+        handleSingleEvent(
+            event,
+            !hasNextEvent && !eventBatch.getWorkflowTaskCompletedEvent().isPresent(),
+            iterator.hasNext() || hasNextEvent);
```

Comment:
I'd suggest lifting these bool operations up into named bindings to make it clear what they're for